src/sparseml/pytorch/sparsification/pruning/mask_creator.py

```diff
@@ -583,17 +584,116 @@ def _get_blocked_tens_shape_and_validate(
+        target: Union[float, List[float]],
+        global_sparsity: bool = False,
+    ) -> List[Tensor]:
+        """
+        :param tensors: list of tensors to calculate a masks based on their contained
+            values
+        :param target: the desired sparsity (decimal fraction of zeros) to reach
+            within the mask or other float target value to base sparsity masks on.
+            Can also be a list where each element is a target for a tensor in the same
+            position in the tensor list. The target value must be within 1e-2 of the
+            effective sparsity of the N:M ratio.
+        :param global_sparsity: typically used to determine pruning masks globally.
+            Not used here because global sparsity doesn't apply to N:M pruning
+        :return: list of masks (0.0 for values that are masked, 1.0 for values that are
+            unmasked) calculated from the tensors such that the desired number of zeros
+            matches the sparsity.
+        """
+        nm_sparsity = 1 - (self._N / self._M)
+        if (isinstance(target, float) and target == 0.0) or (
+            isinstance(target, List) and all([sparsity == 0.0 for sparsity in target])
+        ):
+            return [torch.ones_like(tensor) for tensor in tensors]
+        if (isinstance(target, float) and abs(nm_sparsity - target) > 1e-2) or (
+            isinstance(target, List)
+            and any([abs(nm_sparsity - sparsity) > 1e-2 for sparsity in target])
+        ):
+            raise ValueError(
+                "Sparsity must match N:M ratio. e.g. if using '3:4' then sparsity "
+                "should be set to 0.25"
+            )
+
+        masks = []
+        for tensor in tensors:
+            if tensor.numel() % self._M != 0:
+                raise ValueError(
+                    f"Tensor of size {tensor.shape} can't be evenly divided into "
+                    f"{self._M} groups"
+                )
+            original_tensor = tensor.clone()
+            num_groups = tensor.numel() // self._M
+            if len(tensor.shape) == 4:
+                # N:M sparsity for convolutional layers
+                tensor_temp = (
+                    tensor.detach()
+                    .abs()
+                    .permute(0, 2, 3, 1)
+                    .reshape(num_groups, self._M)
+                )
+                index = torch.argsort(tensor_temp, dim=1)[:, : int(self._M - self._N)]
+                w_b = torch.ones(tensor_temp.shape, device=tensor_temp.device)
+                masks.append(
+                    w_b.scatter_(dim=1, index=index, value=0).reshape(
+                        original_tensor.permute(0, 2, 3, 1).shape
+                    )
+                )
+            elif len(tensor.shape) == 2:
+                # N:M sparsity for linear layers
+                tensor_temp = tensor.detach().abs().reshape(num_groups, self._M)
+                index = torch.argsort(tensor_temp, dim=1)[:, : int(self._M - self._N)]
```

Comment:
If you see fit, perhaps it would make sense to make the code a bit compact here (e.g. L667 and L657 are duplicates). If you think that it would hurt readability, you may disregard this comment.

Reply:
I think it's a good callout, but I'll opt for keeping it as is. Since `index` depends on `tensor_temp` and those are different between the two versions.